lib/Monads/trace/Trace_No_Trace.thy

```diff
@@ -44,34 +44,318 @@ lemma no_trace_is_triple[wp_trip]:
   "no_trace (put s)"
+  by (simp add: no_trace_def put_def)
+
+lemma no_trace_modify[no_trace_terminal]:
   "no_trace (modify f)"
+  by (wpsimp simp: modify_def wp: no_trace_terminal)
+
+lemma no_trace_return[no_trace_terminal]:
   "no_trace (return v)"
+  by (simp add: no_trace_def return_def)
+
+lemma no_trace_fail[no_trace_terminal]:
   "no_trace fail"
-  by (simp_all add: no_trace_def get_def put_def modify_def bind_def
-                    return_def select_def fail_def)
+  by (simp add: no_trace_def fail_def)
+
+lemma no_trace_returnOk[no_trace_terminal]:
+  "no_trace (returnOk v)"
+  by (simp add: returnOk_def no_trace_terminal)
 
-lemma no_trace_select:
+lemma no_trace_select[no_trace_terminal]:
   "no_trace (select S)"
   by (clarsimp simp add: no_trace_def select_def)
 
-lemma no_trace_bind:
-  "no_trace f \<Longrightarrow> \<forall>rv. no_trace (g rv)
-    \<Longrightarrow> no_trace (do rv \<leftarrow> f; g rv od)"
-  apply (subst no_trace_def)
-  apply (clarsimp simp add: bind_def split: tmres.split_asm;
-    auto dest: no_traceD[rotated])
-  done
-
-lemma no_trace_extra:
+lemma no_trace_gets[no_trace_terminal]:
   "no_trace (gets f)"
+  by (wpsimp simp: gets_def wp: no_trace_terminal)
+
+lemma no_trace_assert[no_trace_terminal]:
   "no_trace (assert P)"
+  by (simp add: assert_def no_trace_terminal)
+
+lemma no_trace_case_option[no_trace_cond]:
+  assumes f: "no_trace f"
+  assumes g: "\<And>x. no_trace (g x)"
+  shows "no_trace (case_option f g x)"
+  by (clarsimp simp: f g split: option.split)
+
+lemma no_trace_assert_opt[no_trace_terminal]:
   "no_trace (assert_opt Q)"
-  by (simp_all add: gets_def assert_def assert_opt_def no_trace_bind no_trace_prim
-             split: option.split)
+  by (simp add: assert_opt_def no_trace_terminal no_trace_cond)
+
+lemma no_trace_alt[no_trace_cond]:
+  "\<lbrakk> no_trace f; no_trace g \<rbrakk> \<Longrightarrow> no_trace (f \<sqinter> g)"
+  apply (subst no_trace_def)
+  apply (clarsimp simp add: alternative_def;
+         fastforce dest: no_traceD)
+  done
+
+lemma no_trace_when[no_trace_cond]:
+  "\<lbrakk>P \<Longrightarrow> no_trace f\<rbrakk> \<Longrightarrow> no_trace (when P f)"
+  by (simp add: when_def no_trace_terminal)
+
+lemma no_trace_unless[no_trace_cond]:
+  "\<lbrakk>\<not>P \<Longrightarrow> no_trace f\<rbrakk> \<Longrightarrow> no_trace (unless P f)"
+  by (simp add: unless_def when_def no_trace_terminal)
+
+lemma no_trace_if[no_trace_cond]:
+  "\<lbrakk> P \<Longrightarrow> no_trace f; \<not>P \<Longrightarrow> no_trace g \<rbrakk> \<Longrightarrow> no_trace (if P then f else g)"
+  by simp
+
+lemma no_trace_apply[no_trace_cond]:
+  "no_trace (f (g x)) \<Longrightarrow> no_trace (f $ g x)"
+  by simp
+
+\<comment> \<open>FIXME: make proof nicer\<close>
+lemma no_trace_liftM_eq[simp]:
+  "no_trace (liftM f m) = no_trace m"
+  apply (clarsimp simp: no_trace_def bind_def liftM_def return_def split_def
+                 split: tmres.split_asm)
```

Comment:
Could this proof work by combining `no_trace` results about `bind` and `return` instead of unfolding? (Doesn't have to be in this PR)

Reply:
Unfortunately not, at least not without new `bind` and `return` results. That would work for the following `no_trace_liftM` lemma but not for this one with its equality.

To be honest, I probably would have skipped this equality and just done `no_trace_liftM`, except for the fact that I was copying a lot of the structure of this file from `Trace_No_Fail` and `Trace_Empty_Fail` which both have `liftM_eq` equivalents.

Reply:
Ok, no worries.